### PR TITLE
datanames slot in `teal_transform_module` (ignore `@datanames`)

### PR DESCRIPTION
**Pull request by gogonzo**
Alternative solution to https://github.com/insightsengineering/teal/pull/1327

```
#' @param datanames (`character`) A vector with `datanames` that are relevant for the item. The
#'   filter panel will automatically update the shown filters to include only
#'   filters in the listed datasets. `NULL` will hide the filter panel,
#'   and the keyword `"all"` will show filters of all datasets. `datanames` also determines
#'   a subset of datasets which are appended to the `data` argument in server function.
```

If you look at the `?module` documentation of `datanames` argument, there is nothing about "if datanames is 'all' then all datasets from @datanames will be taken". It is very hard to explain as `datanames` is not only a `module` property, which makes it difficult to describe in a documentation. However, in the vignette we have it described in details [see here](https://insightsengineering.github.io/teal/latest-tag/articles/including-data-in-teal-applications.html#teal_data-properties)

Here we have a following rules:
1. Module's server get's always `module$datanames` if specified. 
    - When transform argument  is specified in the `module(transform)` then module appends `transform$datanames` to `module$datanames` to always keep `module$datanames` as a single point of truth. `transform` is always added as an argument of `module()` function, so resolving datanames there makes sense and is the most optimal. 
2. When `module$datanames = "all"` then all objects from transformed `data` is passed to the module.
    - When data contains datasets which shouldn't be displayed for a given module then one should `module |> set_datanames()` to limit displayed datasets
    

    
 **Advantage:**
- single point of truth (`module$datanames`) 
- don't need to take care of `data@datanames`
- can re-set `datanames` for the whole group of modules if needed 
 
 **Disadvantages:**
 - If re-set is applied for all modules `modules(mod1, mod2, mod3) |> set_datanames()` then each module's datanames will be re-set even if their `datanames` have been set to what is exactly needed. Possible solution is to `append_datanames()` instead of `set_datanames()`, WDYT?
 
 <details>
 <summary>app</summary>
 
 ```r
pkgload::load_all("teal")

trans <- list(
  anl_w_datanames = teal_transform_module(
    label = "ANL with datanames",
    ui = function(id) {
      ns <- NS(id)
      tagList(
        div("This transformer adds ANL based on specified ADSL, ADTTE"),
        numericInput(ns("obs"), "Number of subjects", value = 400, min = 0, max = 400)
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(),
            {
              ANL <- dplyr::inner_join(
                head(ADSL, nobs),
                ADTTE[c("STUDYID", "USUBJID", setdiff(colnames(ADTTE), colnames(ADSL)))],
                by = c("USUBJID", "STUDYID")
              )
            },
            nobs = input$obs
          )
        })
      })
    },
    datanames = "ADTTE"
  ),
  anl_wout_datanames = teal_transform_module(
    label = "ANL without datanames",
    ui = function(id) {
      ns <- NS(id)
      tagList(
        div("This transformer adds ANL based on unspecified ADSL, ADTTE"),
        numericInput(ns("obs"), "Number of subjects", value = 400, min = 0, max = 400)
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(),
            {
              ANL <- dplyr::inner_join(
                head(ADSL, nobs),
                ADTTE[c("STUDYID", "USUBJID", setdiff(colnames(ADTTE), colnames(ADSL)))],
                by = c("USUBJID", "STUDYID")
              )
            },
            nobs = input$obs
          )
        })
      })
    }
  ),
  adsl_w_datanames = teal_transform_module(
    label = "modify ADSL",
    ui = function(id) {
      ns <- NS(id)
      tagList(
        div("This transformer modifies ADSL based on specified ADTTE")
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(),
            {
              ADTTE_summary <- ADTTE |>
                dplyr::group_by(STUDYID, USUBJID) |>
                dplyr::summarize(PARAMCD_AVAL = paste(paste(PARAMCD, "-", AVAL), collapse = "; "))
              ADSL <- dplyr::left_join(ADSL, ADTTE_summary)
            },
            nobs = input$obs
          )
        })
      })
    },
    datanames = c("ADSL", "ADTTE")
  ),
  adsl_wout_datanames = teal_transform_module(
    label = "modify ADSL",
    ui = function(id) {
      ns <- NS(id)
      tagList(
        div("This transformer modifies ADSL based on unspecified ADTTE")
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(),
            {
              ADTTE_summary <- ADTTE |>
                dplyr::group_by(STUDYID, USUBJID) |>
                dplyr::summarize(PARAMCD_AVAL = paste(sprintf("%s - %.2f", PARAMCD, AVAL), collapse = "; "))
              ADSL <- dplyr::left_join(ADSL, ADTTE_summary)
            },
            nobs = input$obs
          )
        })
      })
    },
    datanames = "ADSL"
  )
)

data <- teal_data_module(
  once = FALSE,
  ui = function(id) {
    ns <- NS(id)
    tagList(
      numericInput(ns("obs"), "Number of observations to show", 1000),
      actionButton(ns("submit"), label = "Submit")
    )
  },
  server = function(id, ...) {
    moduleServer(id, function(input, output, session) {
      logger::log_trace("example_module_transform2 initializing.")
      eventReactive(input$submit, {
        data <- within(data = teal_data(), n = as.numeric(input$obs), {
          logger::log_trace("Loading data")
          z <- NULL
          ADSL <- head(teal.data::rADSL, n = n)
          ADTTE <- teal.data::rADTTE
          iris <- iris
          iris_raw <- iris
          CO2 <- CO2
          factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
          CO2[factors] <- lapply(CO2[factors], as.character)
        })
        join_keys(data) <- default_cdisc_join_keys[c("ADSL", "ADTTE")]
        teal.data::datanames(data) <- c("CO2", "ADTTE", "iris_raw", "iris", "ADSL")
        data
      })
    })
  }
)

mods <- modules(
  modules(
    label = "all",
    example_module("all", datanames = "all") |>
      set_datanames("ADSL"), # this will be ignored - set_datanames on the modules level overwrites all items
    example_module("null", datanames = NULL)
  ) |> set_datanames(c("ADSL", "ADTTE", "iris", "iris_raw", "CO2")),
  modules(
    label = "modules datanames",
    example_module("mod-1"),
    example_module("mod-2")
  ) |> set_datanames("ADTTE"), # should include also ADSL (parent)
  modules(
    label = "specified",
    example_module("adsl", datanames = "ADSL"),
    example_module("adsl+adtte", datanames = c("ADSL", "ADTTE")),
    example_module("adtte+parent", datanames = "ADTTE")
  ),
  modules(
    label = "transforms",
    example_module("anl - transform w/ datanames", dataname = "ANL", transformers = trans["anl_w_datanames"]),
    example_module("anl - transform w/o datanames", dataname = "ANL", transformers = trans["anl_wout_datanames"]),
    example_module("adsl - transform w/ datanames", dataname = "ADSL", transformers = trans["adsl_w_datanames"]),
    example_module("adsl - transform w/o datanames", dataname = "ADSL", transformers = trans["adsl_wout_datanames"]),
    example_module("all w/o datanames", dataname = "all", transformers = trans["anl_wout_datanames"]),
    example_module("all reset datanames", dataname = "all", transformers = trans["anl_wout_datanames"]) |> set_datanames(c("ADSL", "ADTTE", "ANL"))
  ),
  modules(
    label = "missings",
    example_module("inexisting+existing", datanames = c("ADTTE", "inexisting")),
    example_module("inexisting", datanames = "inexisting")
  ),
  reporter_previewer_module()
)

slices <- teal_slices(
  teal_slice("ADSL", "SEX"),
  teal_slice("ADSL", "AGE", selected = c(18L, 65L)),
  teal_slice("ADTTE", "PARAMCD", selected = "CRSD"),
  include_varnames = list(
    ADSL = c("SEX", "AGE")
  )
)

app <- teal::init(
  data = data,
  modules = mods,
  filter = slices
)

runApp(app)
 ```
 </details>

### CODE REVIEW

**Comment by donyunardi** [comment]
Thanks for this alternative solution.

To better illustrate the different scenarios, I've created a table based on the example app that @gogonzo provided. Just to clarify, the `data` passed to the module contains values in the `data$datanames` slot.

| # | module$datanames | use transform? | set transform$datanames? | \|> set_datanames()? | module gets | 
| :--: | :--------: | :--------: | :--------: | :--------: | :--------: | 
| 1 | ADSL | ❌ | ❌ | ❌ | ADSL |
| 2 | ADSL | ✅ | ANL | ❌ | ADSL, ANL |
| 3 | ADTTE | ❌ | ❌ | ❌ | ADTTE, ADSL (parent) |
| 4 | ANL | ✅ | ❌ | ❌ | ANL |
| 5 | ANL | ✅ | ADSL | ❌ | ANL, ADSL |
| 6 | ANL | ✅ | ADTTE | ❌ | ANL, ADTTE, ADSL (parent) 
| 7 | all | ❌ | ❌ | ❌ | all objects in `data$env` |
| 8 | all | ✅ | ADSL | ❌ | all objects in `data$env` |
| 9 | all | ✅ | ANL | ❌ | all objects in `data$env` |
| 10 | NULL | ❌ | ❌ | ❌ | all objects in `data$datanames` |
| 11 | NULL | ✅ | ❌ | ❌ | all objects in `data$env` |
| 12 | NULL | ✅ | ANL | ❌ | ANL |
| 13 | NULL | ✅ | ADSL | ❌ | ADSL |
| 14 | NULL | ✅ | ADTTE | ❌ | ADTTE, ADSL |
| 15 | NULL | ✅ | ❌ | ANL | ANL |
| 16 | all | ✅ | ❌ | ANL | ANL |
| 17 | ADSL | ✅ | ANL | ADSL | ADSL|
| 18 | ADSL | ✅ | ❌ | ANL | ANL |
| 19 | ANL | ✅ | ❌ | ADSL | ADSL |

We can use this as reference if needed.

**Comment by donyunardi** [comment]
## What I like

> single point of truth (module$datanames)
> don't need to take care of data@datanames

This makes sense to me. The transform module is designed to be local to the teal module, and `module$datanames` is always intended to either retrieve or limit the data objects pulled from `data`.

> module appends transform$datanames to module$datanames to always keep module$datanames as a single point of truth.

My understanding is that the purpose of `datanames` remains consistent for both the transform and the module, which is to either retrieve or limit the data objects pulled from `data`. I like this as well and it reflected nicely on the scenario table.

## `set_dataname()`
I have some reservations about using `set_datanames`. While I agree it would simplify things, I wonder if it's too brute of an approach for specifying `datanames`. For instance, if `set_datanames` ultimately resets the datanames and modules only access those specified by `set_datanames`, why do we still need the `datanames` argument in `modules$datanames`? 

If the motivation behind `set_datanames` is to handle cases where `modules$datanames` is set to 'all' or 'NULL,' should we limit the use of `set_datanames` to just these scenarios?

## `data$datanames` still needed?
From what I've observed, there seems to be no need to set `data$datanames` during the creation of `teal_data()`. For example, if I omit `CO2` from `datanames` during the creation of a `teal_data_module`, I can still access this data by setting `module$datanames = 'CO2'`. Does this mean users can skip defining `data$datanames` altogether? I don't mind giving users one less thing to do, but I'm wondering if there's another impact we're not considering.

**Comment by gogonzo** [comment]
> If the motivation behind set_datanames is to handle cases where modules$datanames is set to 'all' or 'NULL', should we limit the use of set_datanames to just these scenarios?

Good point, thanks! I think so that `set_datanames()` should work only if `datanames` are `"all"` or `NULL`.

> From what I've observed, there seems to be no need to set data$datanames during the creation of teal_data(). For example, if I omit CO2 from datanames during the creation of a teal_data_module, I can still access this data by setting module$datanames = 'CO2'. 

Yes, that's true. `module$datanames` always had and has a priority over `@datanames`

> Does this mean users can skip defining data@datanames altogether? I don't mind giving users one less thing to do, but I'm wondering if there's another impact we're not considering.

Yes, app-developers can skip `data@datanames` and depend only on `module$datanames`. At this moment we can do following in init:

```r
init <- ...

  if (!is.null(teal.data::datanames(data))) {
    # deprecation message someday
    modules <- set_datanames(teal.data::datanames(data))
  }
  
}
```

**Comment by m7pr** [comment]
Some tests are failining and it looks like the order of items in `data_summary` is different now

**Comment by kumamiao** [comment]
I like that we have a single point of truth `module$datanames`, and agree that `set_datanames()` should only be used when `datanames` are `"all"` or `NULL`.

I actually do like that app-developers can skip `data@datanames`, where it now seems like the functionality of setting `data@datanames` can be achieved by setting `module$datanames`.

File: R/modules.R [hunk 1]
```diff
@@ -39,25 +42,39 @@
...
+#' When `datanames` is set to  `"all"`, all available datasets are considered as relevant for the module.
+#' Sometimes, `data` handed over to `teal` application might contain datasets which are irrelevant for the module
+#' if it has a `datanames` property set to `"all". These objects usually are:
+#' - proxy variables used to modify column etc.
+#' - modified copies of datasets used to create a final dataset.
+#' - connection objects
+#' In order to disable these dataset from showing up in a module one should use `set_datanames` function on the
+#' [module] or [modules()] to set `datanames` from `"all"` to some specific names. Trying to change `datanames`
+#' which are not `"all"` using `set_datanames` will be ignored with warning.
+#' Datasets which name starts with `.` are ignored when `datanames` is set to `"all"`.
```
Comment (gogonzo) [inline R/modules.R:75]: > I like that we have a single point of truth `module$datanames`, and agree that `set_datanames()` should only be used when `datanames` are `"all"` or `NULL`.

@kumamiao additionally one can make `.<object name>` in order to ignore it.

**Review by donyunardi — APPROVED** [review]
Minor sentence update/suggestion.
Everything else looks good. 

Thanks for the thinking hard on this!

**Pull request merged**